### PR TITLE
Simplify VJP/JVP/Grad/Jacobian frontend

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**
Simplify the unflatten of VJP/JVP/grad/Jacobian and covers more edge cases.

### CODE REVIEW

**Comment by josh146** [comment]
This will be really nice to have in 😎 

@rmoyard I think a lot of docstrings/codeblocks (especially the quickstart, sharp bits page, etc.) will likely also need to be updated 🤔

**Comment by rmoyard** [comment]
@josh146 I am double checking the docs

**Comment by rmoyard** [comment]
I have found some changes in the quickstart and sharp bits, and updated them

**Review by dime10 — CHANGES_REQUESTED** [review]
Looks great, thanks for going over those PRs again 👍

File: frontend/catalyst/pennylane_extensions.py [hunk 1]
```diff
@@ -947,19 +943,17 @@ def check_is_iterable(x, hint):
...
-        else:
-            func_res = results[: len(jaxpr.out_avals)]
-            vjps = results[len(jaxpr.out_avals) :]
-            results = tuple([*func_res, tuple(vjps)])
+        func_res = results[: len(jaxpr.out_avals)]
+        vjps = results[len(jaxpr.out_avals) :]
+        func_res = tree_unflatten(out_tree, func_res)
+        # We do not change the shape of the VJP as it is the same as the parameters not
+        # as the output
+        results = tuple([func_res, tuple(vjps)])
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:955]: No currently parameters are not pytrees, so it will be done in the upcoming PR when I add support for pytrees params

File: frontend/catalyst/pennylane_extensions.py [hunk 2]
```diff
@@ -947,19 +943,17 @@ def check_is_iterable(x, hint):
 
         jaxpr, out_tree = _make_jaxpr_check_differentiable(fn, grad_params, *params)
 
+        cotangents, _ = tree_flatten(cotangents)
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:946]: Yes I think it is part of another ticket to check the shape of cotangent and tangents

File: frontend/catalyst/pennylane_extensions.py [hunk 3]
```diff
@@ -873,15 +870,14 @@ def check_is_iterable(x, hint):
...
-            results = tuple([func_res, jvps])
-        else:
-            results = tuple(results)
+        midpoint = len(results) // 2
+        func_res = results[:midpoint]
+        jvps = results[midpoint:]
+
+        func_res = tree_unflatten(out_tree, func_res)
+        jvps = tree_unflatten(out_tree, jvps)
+        results = tuple([func_res, jvps])
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:879]: No no worries :+1:

File: frontend/catalyst/pennylane_extensions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:955]: I mean you already flatten the cotangents, but in that case I would mark it as a TODO here that the output shape is to be updated.

**Comment by rmoyard** [comment]
@dime10 I have added a note in each function that we support function that returns any pytree-like shape

File: frontend/catalyst/pennylane_extensions.py [hunk 4]
```diff
@@ -794,7 +791,8 @@ def jvp(f: DifferentiableLike, params, tangents, *, method=None, h=None, argnum=
 
     This function allows the Jacobian-vector Product of a hybrid quantum-classical function to be
     computed within the compiled program. Outside of a compiled function, this function will simply
-    dispatch to its JAX counterpart ``jax.jvp``.
+    dispatch to its JAX counterpart ``jax.jvp``. The function ``f`` can return any pytree-like
+    shape.
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:795]: I think my point was more about checking the type signature and description of arguments and results of the gradient functions. For example looking at this one:
```
    Returns (Tuple[Array]):
        Return values of ``f`` paired with the JVP values.
```
I guess the description of it is already super generic, so it could still be considered "correct" but I think it would be better if we actually described what the output looks like.

The return type (which is actually incorrectly formatted here to be picked up by sphinx) is also outdated now since we are no longer restricted to Arrays. I'm not sure about the best type to use in its place, JAX seems to just use `Tuple[Any]`.

Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:794]: I have change the signature to tuple(any) and fixed the format issue in VJP and JVP, let me know if that works on your side.

**Review by dime10 — APPROVED** [review]
Looks good on my side 👍

File: doc/dev/quick_start.rst [hunk 5]
```diff
@@ -636,8 +636,8 @@ decorator to compute Jacobian matrices of general hybrid functions with multiple
         return g(x)
 
 >>> workflow(jnp.array([2.0, 1.0]))
-array([[-1.32116540e-07,  1.33781874e-07],
-       [-4.20735506e-01,  4.20735506e-01]])
+array([[ 1.74393425e-16  4.54648713e-01]
+       [-1.74393425e-16 -4.54648713e-01]])
```
Comment (josh146) [inline doc/dev/quick_start.rst:640]: do we expect the gradient to change 😅

File: doc/dev/quick_start.rst [hunk 6]
```diff
@@ -678,7 +678,7 @@ the value of ``value_and_grad`` argument. To optimize params iteratively, you la
     def workflow():
         def gd_fun(param):
             diff = grad(circuit, argnum=0)
-            return circuit(param), diff(param)[0]
+            return circuit(param), diff(param)
```
Comment (josh146) [inline doc/dev/quick_start.rst:681]: 🙌

File: doc/dev/quick_start.rst (hunk 5 above)
Comment (rmoyard) [reply doc/dev/quick_start.rst:640]: I have copied the wrong one normally there is just a transposition, sorry for that. And this was introduced in a previous PR. Open a Pr shortly.

```
[[ 3.48786850e-16 -4.20735492e-01]
 [-8.71967125e-17  4.20735492e-01]]
```
